### PR TITLE
Allow JWT token auth in binary protocol

### PR DESCRIPTION
**Pull request by fantix**
Cherry-pick #4587, #4703, #4708 and #4829, and add support for both old and new CLI

Tested:

- [x] CLI 2.2.6+7eabbf9 `edgedb ui`
- [x] CLI 2.3.0 dev `edgedb ui`
- [x] UI password login (uses server-generated token)

### CODE REVIEW

File: edb/server/protocol/binary.pyx [hunk 1]
```diff
@@ -667,22 +686,18 @@ cdef class EdgeConnection(frontend.FrontendConnection):
             ) from None
 
         if not claims.get(f"{namespace}.any_role"):
+            assert not is_legacy  # legacy CLI will only generate any_role JWTs
             token_roles = claims.get(f"{namespace}.roles")
-            if not isinstance(token_roles, dict):
+            if not isinstance(token_roles, list):
```
Comment (elprans) [inline edb/server/protocol/binary.pyx:691]: Legacy tokens would still have a dict here, no?

Comment (fantix) [reply edb/server/protocol/binary.pyx:691]: Yes, that's why I put an assertion above - I think we could also do:

```suggestion
            if not isinstance(token_roles, dict if is_legacy else list):
```

Comment (elprans) [reply edb/server/protocol/binary.pyx:691]: Well, yes, otherwise it will crash on a legacy token.

**Review by msullivan — APPROVED** [review]

**Comment by AnsonHwang86** [comment]
@raddevon Can add document to this? Without document, I don't know how to use JWT in edgedb. thanks.

**Pull request merged**